File: hermes/src/main/java/org/eclipse/paho/android/service/contract/DefMqttEventActionListener.java
package org.eclipse.paho.android.service.contract;

import org.eclipse.paho.android.service.event.MqttAction;
import org.eclipse.paho.android.service.event.MqttEvent;
import org.eclipse.paho.client.mqttv3.IMqttActionListener;
import org.eclipse.paho.client.mqttv3.IMqttToken;
import org.greenrobot.eventbus.EventBus;

/**
 * 默认实现MQTT消息监听接口
 * @author zhouL
 * @date 2023/10/9
 */
public class DefMqttEventActionListener implements IMqttActionListener {

    protected final String clientKey;

    protected final MqttAction action;

    protected final String errorMsg;

    public DefMqttEventActionListener(String clientKey, MqttAction action, String errorMsg) {
        this.clientKey = clientKey;
        this.action = action;
        this.errorMsg = errorMsg;
    }

    @Override
    public void onSuccess(IMqttToken asyncActionToken) {
        EventBus.getDefault().post(MqttEvent.createSuccess(clientKey, action));
    }

    @Override
    public void onFailure(IMqttToken asyncActionToken, Throwable exception) {
        exception.printStackTrace();
        EventBus.getDefault().post(MqttEvent.createFail(clientKey, action, errorMsg, exception));
    }
}

File: hermes/src/main/java/com/lodz/android/hermes/mqtt/base/contract/DefMqttEventActionListener.kt
package com.lodz.android.hermes.mqtt.base.contract

import com.lodz.android.hermes.mqtt.base.bean.eun.MqttAction
import com.lodz.android.hermes.mqtt.base.bean.event.MqttEvent
import org.eclipse.paho.client.mqttv3.IMqttActionListener
import org.eclipse.paho.client.mqttv3.IMqttToken
import org.greenrobot.eventbus.EventBus
import java.lang.RuntimeException

/**
 * 默认实现MQTT消息监听接口
 * @author zhouL
 * @date 2023/10/12
 */
open class DefMqttEventActionListener(
    protected val clientKey: String,
    protected val action: MqttAction,
    protected val errorMsg: String
) : IMqttActionListener {

    override fun onSuccess(asyncActionToken: IMqttToken?) {
        EventBus.getDefault().post(MqttEvent.createSuccess(clientKey, action))
    }

    override fun onFailure(asyncActionToken: IMqttToken?, exception: Throwable?) {
        exception?.printStackTrace()
        EventBus.getDefault().post(MqttEvent.createFail(clientKey, action, errorMsg, getThrowable(exception)))
    }

    protected fun getThrowable(t: Throwable?): Throwable = t ?: RuntimeException()
}